base/reducedim.jl

```diff
@@ -3,70 +3,58 @@
 ## Functions to compute the reduced shape
 
 # for reductions that expand 0 dims to 1
-reduced_dims(a::AbstractArray, region) = reduced_dims(size(a), region)
+reduced_dims(a::AbstractArray, region) = reduced_dims(indices(a), region)
 
 # for reductions that keep 0 dims as 0
-reduced_dims0(a::AbstractArray, region) = reduced_dims0(size(a), region)
-
-function reduced_dims{N}(siz::NTuple{N,Int}, d::Int, rd::Int)
-    if d < 1
-        throw(ArgumentError("dimension must be ≥ 1, got $d"))
-    elseif d == 1
-        return tuple(rd, siz[d+1:N]...)::typeof(siz)
-    elseif 1 < d < N
-        return tuple(siz[1:d-1]..., rd, siz[d+1:N]...)::typeof(siz)
-    elseif d == N
-        return tuple(siz[1:N-1]..., rd)::typeof(siz)
+reduced_dims0(a::AbstractArray, region) = reduced_dims0(indices(a), region)
+
+function reduced_dims{N}(inds::Indices{N}, d::Int, rd::AbstractUnitRange)
+    d < 1 && throw(ArgumentError("dimension must be ≥ 1, got $d"))
+    if d == 1
+        return (oftype(inds[1], rd), tail(inds)...)
+    elseif 1 < d <= N
+        return tuple(inds[1:d-1]..., oftype(inds[d], rd), inds[d+1:N]...)::typeof(inds)
     else
-        return siz
+        return inds
     end
 end
-reduced_dims{N}(siz::NTuple{N,Int}, d::Int) = reduced_dims(siz, d, 1)
+reduced_dims{N}(inds::Indices{N}, d::Int) = reduced_dims(inds, d, OneTo(1))
 
-function reduced_dims0{N}(siz::NTuple{N,Int}, d::Int)
-    if d < 1
-        throw(ArgumentError("dimension must be ≥ 1, got $d"))
-    elseif d <= N
-        return reduced_dims(siz, d, (siz[d] == 0 ? 0 : 1))
+function reduced_dims0{N}(inds::Indices{N}, d::Int)
+    d < 1 && throw(ArgumentError("dimension must be ≥ 1, got $d"))
+    if d <= N
+        return reduced_dims(inds, d, (inds[d] == OneTo(0) ? OneTo(0) : OneTo(1)))
     else
-        return siz
+        return inds
     end
 end
 
-function reduced_dims{N}(siz::NTuple{N,Int}, region)
-    rsiz = [siz...]
+function reduced_dims{N}(inds::Indices{N}, region)
+    rinds = [inds...]
     for i in region
         isa(i, Integer) || throw(ArgumentError("reduced dimension(s) must be integers"))
-        d = convert(Int, i)::Int
+        d = Int(i)
         if d < 1
             throw(ArgumentError("region dimension(s) must be ≥ 1, got $d"))
         elseif d <= N
-            rsiz[d] = 1
+            rinds[d] = oftype(rinds[d], OneTo(1))
         end
     end
-    tuple(rsiz...)::typeof(siz)
+    tuple(rinds...)::typeof(inds)
 end
 
-function reduced_dims0{N}(siz::NTuple{N,Int}, region)
-    rsiz = [siz...]
+function reduced_dims0{N}(inds::Indices{N}, region)
+    rinds = [inds...]
```

Comment:
It turns out that splatting is more efficient, at least for the couple of `N`s I tried.